File: Assignment-3/oct03/c/wada.c
#include "rk4.h"
#include "pendulum.h"
#include <stdio.h>

//  The maximum period of any orbit under consideration by this program.
static size_t maxperiod = 2;

//  Here is one possible way to collect all the information relevant to
//  tracking the periodic orbits.  In this format, only one representative
//  of each periodic orbit is stored in PT; you have to iterate the map
//  to check whether a candidate point comes within EPSILON of the stored orbit.

struct fixedpoint {
   real_t pt[3][2];  // the list of periodic points
   real_t epsilon;   // convergence criterion
   size_t period[3];  // their respective periods
   size_t maxiter;  // maximum number of map iterations
   size_t nfixed;  // the number of fixed points in the list (i.e., 3)
};

// The fixed points whose basins we will calculate.  One representative
// of each point is stored.  This is a static initialization; the assignments
// are done at compile and link time.

static struct fixedpoint fp = {
   {{-2.407661292643E+00,  6.773267621896E-01},  // period 1
    {-6.099150484926E-01,  1.677463819037E+00},  // period 2
    {-7.218820695570E-01, -4.620944521247E-01}  // period 2
   },  // pt
   1.0E-09,  //  required approach distance to confirm convergence
   {1, 2, 2}, // the respecitve periods
   400, // maxiter
   3 // nfixed
};

//  This is one possible way to represent the grid to be used for computing
//  the basin boundary.  We assume a square grid here but it's straightforward
//  to extend the structure to accommodate a rectangular one.

struct grid {
   real_t xmin;  // angular position limits
   real_t xmax;
   real_t ymin;  // angular velocity limits
   real_t ymax;
   size_t resolution;  // number of grid points in each direction
};

// The grid used to generate the basin.  Unfortunately, in Standard C,
// you cannot declare PI and MINUSPI as CONST quantities, which is less
// error-prone than simple macros.  Expressions like -PI (where PI is a
// CONST variable) are permitted in GNU C as an extension, but not in Intel C.
// For portability reasons, it is best to avoid language extensions.

static struct grid box = {
  MINUSPI, PI,  // xmin, xmax
  MINUSPI, PI,  // ymin, ymax
  400  // resolution
};
//----------------------------------------------------------------------------
// Check whether Y or its orbit contains a point that is already in the
// list of fixed points.  If so, then return an integer value (1, 2, or 3)
// indicating whether it is the first, second, or third point in FIXEDPOINT.
// In other words, if Y lies within fp.epsilon of pt[i], then return i+1,
// i = 0, 1, 2.  Otherwise, return 0.
// You can use any metric here.  The 2-norm in C can be computed with the
// standard library function HYPOT(), but the 0-norm also works as the
// maximum of the absolute values of the differences between any component.
// C99 includes the functions FMAX() and FABS(), which are convenient for this
// purpose.

int
check_point(real_t y[2])
{
   // your code here
   return(0);  // substitute the appropriate value
}
//----------------------------------------------------------------------------
// Compute the basin boundary of each fixed point.  For the (J,K)th grid point,
// compute up to MAXITER iterations of the Poincare map and see where the
// initial condition ends up.  If it lies within EPSILON of one of the fixed
// points (which CHECK_POINT will determine), then mark the basin accordingly.

int*
compute_basin(const struct grid *b)
{
    size_t n = b->resolution;
    int *basin = (int *) calloc(n*n, sizeof(*basin));  // initialized to 0

    if(basin == NULL) return(NULL);  // allocation failure

    // your code here
    return(basin);
}
//----------------------------------------------------------------------------
// Main program.  
// The version below simply allocates an NxN array of integers for the
// basin boundary array, and you will need to do all the subscript arithmetic
// explicitly to compute the offset needed to access the (j,k)th element.
// Alternatively, you can to convert BASIN to an int**,
// allocate N int*'s, then allocate the NxN block and precompute the
// address of each row (see Sept. 26 lecture notes for details).
// In this case, you'll need to modify the FWRITE call that writes out
// the BASIN array.
int
main(int argc, char **argv)
{
    size_t n = box.resolution;
    const char *filename = "basin.dat";
    FILE *outfp = fopen(filename, "wb");   // 'b' for binary
    int *basin;

    if(outfp == NULL) {  // cannot open output file, so halt
       perror(filename);
       return(EXIT_FAILURE);
    }

    basin = compute_basin(&box);
    if(basin == NULL) {  // heap disaster; printing a message is
       return(EXIT_FAILURE);   // unlikely to succeed
    }

    // Write out the results in a binary format compatible with the MATLAB
    // script wada.m for visualization.

    fwrite(&box, sizeof(box.xmin), 4, outfp);  // xmin through ymax
    fwrite(&fp.epsilon, sizeof(fp.epsilon), 1, outfp);
    fwrite(&box.resolution, sizeof(box.resolution), 1, outfp);
    fwrite(basin, sizeof(*basin), n*n, outfp);  // N x N basin boundary
    fclose(outfp);
    free(basin);
    return(EXIT_SUCCESS);
}

File: Assignment-3/oct03/fortran/wada.f90
!  WADA - find the periodic points in a given domain for the
!  forced damped nonlinear pendulum with a fixed set of parameters.
!
      module wada
      use pendulum
      implicit none
!
!  Grid that defines the region over which we compute the basin.
!  The initializations below are the default values assigned to each
!  component when a new instance is created.
!
      type::grid
         real(WP):: xmin=-PI  ! grid limits in X and Y
         real(WP):: xmax=PI
         real(WP):: ymin=-PI
         real(WP):: ymax=PI
         integer:: resolution=400  ! grid resolution in each direction
      end type grid
!
!  Fixed points obtained with the default map parameters.
!  These values are all precomputed by another program.
!
      integer,parameter:: MAXPERIOD=2
      integer,parameter:: MAXPTS=3
      type:: fixedpoints
         real(WP)::pt(2,MAXPTS)=reshape([                    &
           -2.407661292643_WP, 0.6773267621896_WP,           &
           -0.6099150484926_WP,  1.677463819037_WP,          &
           -0.7218820695570_WP, -0.4620944521247_WP], [2, MAXPTS])
         real(WP):: epsilon=1.0e-09_WP  ! distance threshold
         integer:: period(MAXPTS)=[1,2,2]
         integer:: maxiter=400  ! max number of map iterations
         integer:: nfound=MAXPTS
      end type fixedpoints
!
!  The only instance of the FIXEDPOINTS type in this program.
!
      type(fixedpoints), save, private:: fp
      contains
!----------------------------------------------------------------------
      subroutine check_point(y,eps,idx)
!  CHECK_POINT - determine which of the periodic points, if any,
!  Y and its orbit approach.
!  Y : the 2-vector whose orbit is examined.
!  EPS : distance threshold; if the orbit approaches within EPS, then
!   we conclude that the limiting point is the IDXth one.
!  IDX := set to M if the orbit of Y comes within EPS of the Mth fixed point,
!   M=1,2,3.  Otherwise, IDX is set to 0.
!
      real(WP), intent(in):: y(2), eps
      integer, intent(out):: idx

      !  Your code here

      idx = 0  ! set to the appropriate value
      return
      end subroutine check_point
!----------------------------------------------------------------------
      subroutine compute_basin(box,basin)
!  COMPUTE_BASIN - compute the Wada basin boundary.
!
      type(grid), intent(in):: box
      integer, allocatable, intent(out):: basin(:,:)
!
!  Local variables
!
      integer:: n  ! among others that you will need
!
!  The array BASIN holds the fate of each point in the grid.
!  BASIN(J,K) is one of 1, 2, or 3 according as the initial condition
!  (x_j, y_k) tends to fixed point A, B or C (defined as fp%pt(:,1),
!  fp%pt(:,2), and fp%pt(:,3), respectively), or as 0 if the orbit tends
!  to none of them.  The initialization of the array defines each point
!  as none of them, but your code here will made updated determinations.
!
      n = box%resolution
      allocate(basin(n,n))  ! N x N array that defines the fate of each point
      basin = 0
!
!  Compute the basin boundary
!
      return
      end subroutine compute_basin
!----------------------------------------------------------------------
      end module wada
!----------------------------------------------------------------------
! MAIN.
      use wada
      implicit none
      integer, parameter:: IOUT=3  ! output unit number
      integer, allocatable:: basin(:,:)
      type(grid):: box
!
      open(IOUT, file='basin.dat', form='unformatted', access='stream',  &
        status='replace', position='rewind')
      call compute_basin(box,basin)
      write(IOUT) box, basin
      close(IOUT)
      stop
      end

